### DISCUSSION TITLE
Am I doing this right: using context to add systems

### DISCUSSION THREAD

**Discussion by rudiHammad** [skypjack/entt#1031] [Q&A] [answered]
Hello,
as mentioned in my first post in this forum I just started learning Entt and I want to make sure I am doing this properly. I read crash course section about context, didn't find anything specific about how to a add systems, so I thought that context is how to do it?
In the ECS I wrote as a learning exercise I had someting like this `registry->addSystem<MovementSystem>()`. And then I had a `registry->getSystem<MovementSystem>().update(delta)` in my gameloop. 
Translating this to entity this is what I did:

```
// The system
struct MovementSystem
{
public:
    MovementSystem() = default;
    void update(float delta, entt::registry& registry) 
    {
          //....
    }

// In the GameScene intialize method I emplace the system with 
bool GameScene::init()
{    
    //...
    registry.ctx().emplace<MovementSystem>();          // (registry being a private member of GameScene)
    //...
}

// in my GameScene update method I do 
void GameScene::update(float delta)
{
    auto& moveSystem = registry.ctx().get<MovementSystem>();
    moveSystem.update(delta, registry);
}
```

Everything is working as expected, so it is all good. I just want to make sure that adding system to registry through the context is a valid solution, or if there is another more appropriate way.

thanks,
R

**Comment by skypjack** [comment, marked as answer]
I think this is a matter of tastes. In my case, all systems are plain free functions. No need to register them anywhere.
The update function of a scene looks like this:
```cpp
bool my_scene::update(const renderer &renderer) {
    imgui_system(registry);
    rmlui_context->Update();
    // ....
    scene_system(renderer, registry, owner());
    return true;
}
```
However, this isn't the first time I've heard of systems being _registered_ with the registry. So I guees it works too.
I strongly prefer to flip the problem on its head because I want each scene to be able to choose to have one, ten, or a hundred registries as well as none at all and just work without a registry (for example, imagine a credits scene, where it's basically all UI).
A registry-centric design would make it a little harder to achieve, although not impossible. That being said, yeah, it's really a matter of tastes probably. 🤷‍♂️

**Comment by Daedie-git** [comment]
It all comes down to what you want to accomplish. 

Do you want to be able add and remove systems at runtime? Then this approach will not suffice. You might want to consider a separate class that manages systems polymorphically (*)

Are all systems known at compile time? Your approach might work, but is there any benefit to doing this over just declaring your systems on the stack and executing them in the correct order?

Another aspect you might want to think about is whether your want all your systems to be freely accessible from inside any individual system, because that's essentially what you are doing.

My generalized advice would be that the chance is high that doing this is ill-advised. But you might have a strong case for it.


(*) the polymorphic approach might look something like this:

```
class Systems
{
public:
    void update(float delta, entt::registry& registry) override
    {
          //....
    }
    
    template<typename SystemType>
    bool addSystem()
    {
          // check for uniqueness and add the system
          // if you want to enforce that only the registry is allowed to have state, you could add a compile time check here
    }

private:
    std::vector<System*> systems;
};

struct MovementSystem: public System
{
    MovementSystem() = default;
    void update(float delta, entt::registry& registry) override
    {
          //....
    }
};
```

Systems would be an object that lives separate from your registry.

**Reply by rudiHammad** [reply]
Great, thanks a lot for the feed back. I am very new to game development and all I've done so far is a little demo of a space shooter so 
I didn't worry to much about it removing or adding systems. But I will certainly consider what you said.
Thank you.
